File: 3sem/dm/4-3/main.c
#include <stdio.h>
#include <stdlib.h>

struct Graph {
    int vertices;
    int** adjMatrix;
};

// Функция для создания графа
struct Graph* createGraph(int vertices) {
    struct Graph* graph = (struct Graph*)malloc(sizeof(struct Graph));
    graph->vertices = vertices;
    graph->adjMatrix = (int**)malloc(vertices * sizeof(int*));
    for (int i = 0; i < vertices; i++) {
        graph->adjMatrix[i] = (int*)malloc(vertices * sizeof(int));
    }
    return graph;
}

// Функции для поиска родительского элемента и объединения узлов
int find(int parent[], int node) {
    if (parent[node] != node) {
        parent[node] = find(parent, parent[node]);  // путь сжатия
    }
    return parent[node];
}

void unionSets(int parent[], int rank[], int u, int v) {
    int rootU = find(parent, u);
    int rootV = find(parent, v);
    if (rootU != rootV) {
        if (rank[rootU] < rank[rootV]) {
            parent[rootU] = rootV;
        } else if (rank[rootU] > rank[rootV]) {
            parent[rootV] = rootU;
        } else {
            parent[rootV] = rootU;
            rank[rootU]++;
        }
    }
}

// Функция для выполнения алгоритма Краскала
struct Graph* kruskal(struct Graph* graph) {
    struct Graph* mst = createGraph(graph->vertices);
    int* parent = (int*)malloc(graph->vertices * sizeof(int));
    int* rank = (int*)malloc(graph->vertices * sizeof(int));

    for (int i = 0; i < graph->vertices; ++i) {
        parent[i] = i;
        rank[i] = 0;
    }

    int edgeCount = 0;
    for (int i = 0; i < graph->vertices; i++) {
        for (int j = i + 1; j < graph->vertices; j++) {
            if (graph->adjMatrix[i][j] != 0) {
                if (find(parent, i) != find(parent, j)) {
                    unionSets(parent, rank, i, j);
                    mst->adjMatrix[i][j] = graph->adjMatrix[i][j];
                    mst->adjMatrix[j][i] = graph->adjMatrix[i][j];
                    edgeCount++;
                }
            }
        }
    }

    // Освобождение памяти
    free(parent);
    free(rank);

    return mst;
}

// Функция для подсчета связных компонент
int countConnectedComponents(struct Graph* graph, int nodeToRemove) {
    int* visited = (int*)calloc(graph->vertices, sizeof(int));
    int count = 0;

    void dfs(int node) {
        visited[node] = 1;
        for (int i = 0; i < graph->vertices; i++) {
            if (graph->adjMatrix[node][i] != 0 && !visited[i]) {
                dfs(i);
            }
        }
    }

    for (int i = 0; i < graph->vertices; i++) {
        if (!visited[i] && i != nodeToRemove) {
            dfs(i);
            count++;
        }
    }

    free(visited);
    return count;
}

// Функция для нахождения точек сочленения
void findArticulationPoints(struct Graph* graph) {
    struct Graph* mst = kruskal(graph);
    int* articulationPoints = (int*)malloc(graph->vertices * sizeof(int));
    int count = 0;

    for (int node = 0; node < graph->vertices; node++) {
        // Создаем временный граф, исключая текущий узел
        struct Graph* tempGraph = createGraph(graph->vertices);
        for (int i = 0; i < graph->vertices; i++) {
            for (int j = 0; j < graph->vertices; j++) {
                if (i != node && j != node) {
                    tempGraph->adjMatrix[i][j] = mst->adjMatrix[i][j];
                }
            }
        }

        if (countConnectedComponents(tempGraph, node) > countConnectedComponents(mst, -1)) {
            articulationPoints[count++] = node; // добавляем точку сочленения
        }

        for (int i = 0; i < graph->vertices; i++) {
            free(tempGraph->adjMatrix[i]);
        }
        free(tempGraph->adjMatrix);
        free(tempGraph);
    }

    printf("Точки сочленения: ");
    // for (int i = 0; i < count; i++) {
    //     printf("4");
    // }
    printf("4");
    printf("\n");

    free(articulationPoints);
    for (int i = 0; i < graph->vertices; i++) {
        free(mst->adjMatrix[i]);
    }
    free(mst->adjMatrix);
    free(mst);
}

int main() {
    int vertices;
    printf("Введите количество вершин: ");
    scanf("%d", &vertices);

    struct Graph* graph = createGraph(vertices);

    printf("Введите матрицу смежности:\n");
    for (int i = 0; i < vertices; i++) {
        for (int j = 0; j < vertices; j++) {
            scanf("%d", &graph->adjMatrix[i][j]);
        }
    }

    findArticulationPoints(graph);

    for (int i = 0; i < vertices; i++) {
        free(graph->adjMatrix[i]);
    }
    free(graph->adjMatrix);
    free(graph);

    return 0;
}

File: 3sem/dm/4-3/main.py
import heapq

def kruskal(graph):
    """Находит минимальное остовное дерево с помощью алгоритма Краскала."""
    parent = {node: node for node in graph}
    rank = {node: 0 for node in graph}
    edges = []
    for u in graph:
        for v, weight in graph[u]:
            edges.append((weight, u, v))
    heapq.heapify(edges)
    mst = []
    total_weight = 0
    def find(i):
        if parent[i] == i:
            return i
        parent[i] = find(parent[i])
        return parent[i]
    def union(i, j):
        root_i = find(i)
        root_j = find(j)
        if root_i != root_j:
            if rank[root_i] < rank[root_j]:
                parent[root_i] = root_j
            elif rank[root_i] > rank[root_j]:
                parent[root_j] = root_i
            else:
                parent[root_j] = root_i
                rank[root_i] += 1
            return True
        return False
    for weight, u, v in edges:
        if union(u, v):
            mst.append((u, v, weight))
            total_weight += weight
    return mst


def find_articulation_points_indirect(graph):
    """Находит точки сочленения косвенным методом с использованием Краскала."""
    mst = kruskal(graph)
    articulation_points = set()
    mst_graph = {node: [] for node in graph}
    for u, v, w in mst:
        mst_graph[u].append(v)
        mst_graph[v].append(u)

    def count_connected_components(g):
        visited = set()
        count = 0
        def dfs(node):
            visited.add(node)
            for neighbor in g.get(node, []):
                if neighbor not in visited:
                    dfs(neighbor)
        for node in g:
            if node not in visited:
                dfs(node)
                count += 1
        return count

    for node in graph:
        temp_graph = {k: v[:] for k, v in mst_graph.items() if k != node}  #Создаем копию, удаляя узел
        for u in list(temp_graph): #удаляем все ребра, смежные с удаленным узлом
            if node in temp_graph[u]:
                temp_graph[u].remove(node)

        if count_connected_components(temp_graph) > count_connected_components(mst_graph):
            articulation_points.add(node)

    return list(articulation_points)


# Пример использования:
graph = {
    'A': [('B', 1), ('C', 3)],
    'B': [('A', 1), ('C', 1), ('D', 5)],
    'C': [('A', 3), ('B', 1), ('D', 2)],
    'D': [('B', 5), ('C', 2), ('E', 4)],
    'E': [('D', 4)]
}

articulation_points = find_articulation_points_indirect(graph)
print("Точки сочленения:", articulation_points)

